### PR TITLE
[gl] Use primitive from graphics pipeline

### PR DESCRIPTION
**Pull request by grovesNL**
Use the primitive from the graphics pipeline for `Draw` and `DrawIndexed` commands.

### CODE REVIEW

**Comment by msiglreith** [comment]
The change is motivated by carrying over the tessellation control point information over to the draw commands, right? I would prefer to set `gl.PatchParameteri` on pipeline change, if possible, as draw calls occur at a higher frequency than pipeline changes imho. Also, we should do the HAL -> GL conversion once at pipeline creation.

**Comment by grovesNL** [comment]
Agreed, I wasn't sure where to put the tessellation control point information though. Should I add it to `GraphicsPipeline` as `Option<gl::types::GLint>`?

**Comment by grovesNL** [comment]
I updated the commit to perform the conversion during pipeline creation, and store `patch_size` on `GraphicsPipeline` (updated during bind - didn't bother to cache it yet).

File: src/backend/gl/src/command.rs [hunk 1]
```diff
@@ -478,8 +479,20 @@ impl command::RawCommandBuffer<Backend> for RawCommandBuffer {
...
+            primitive,
+            patch_size,
+            ..
+        } = pipeline;
+
+        if self.cache.primitive != Some(primitive) {
+            self.cache.primitive = Some(primitive);
+        }
+
+        if let Some(size) = patch_size {
```
Comment (msiglreith) [inline src/backend/gl/src/command.rs]: I think we can also cache this in the future, but not sure if worth it..

File: src/backend/gl/src/conv.rs [hunk 2]
```diff
@@ -44,3 +44,18 @@ pub fn buffer_usage_to_gl_target(usage: buffer::Usage) -> Option<t::GLenum> {
...
+        Primitive::PointList => gl::POINTS,
+        Primitive::LineList => gl::LINES,
+        Primitive::LineStrip => gl::LINE_STRIP,
+        Primitive::TriangleList => gl::TRIANGLES,
+        Primitive::TriangleStrip => gl::TRIANGLE_STRIP,
+        Primitive::LineListAdjacency => gl::LINES_ADJACENCY,
+        Primitive::LineStripAdjacency => gl::LINE_STRIP_ADJACENCY,
+        Primitive::TriangleListAdjacency => gl::TRIANGLES_ADJACENCY,
+        Primitive::TriangleStripAdjacency => gl::TRIANGLE_STRIP_ADJACENCY,
+        Primitive::PatchList(_) => gl::PATCHES
```
Comment (msiglreith) [inline src/backend/gl/src/conv.rs]: nit: trailing comma

File: src/backend/gl/src/device.rs [hunk 3]
```diff
@@ -347,8 +347,16 @@ impl d::Device<B> for Device {
                     name
                 };
 
+                let patch_size = if let Primitive::PatchList(size) = desc.input_assembler.primitive {
```
Comment (msiglreith) [inline src/backend/gl/src/device.rs]: Let's do a `match` here, looks a bit better/more coherent(?)

File: src/backend/gl/src/queue.rs [hunk 4]
```diff
@@ -463,6 +463,9 @@ impl CommandQueue {
             com::Command::SetDrawColorBuffers(num) => {
                 state::bind_draw_color_buffers(&self.share.context, num);
             }
+            com::Command::SetPatchSize(num) => unsafe {
+                &self.share.context.PatchParameteri(gl::PATCH_VERTICES, num);
```
Comment (msiglreith) [inline src/backend/gl/src/queue.rs]: neat :D

File: src/backend/gl/src/device.rs [hunk 5]
```diff
@@ -347,8 +347,16 @@ impl d::Device<B> for Device {
...
+                    Some(size as _)
+                } else {
+                    None
+                };
+
                 Ok(n::GraphicsPipeline {
                     program,
+                    primitive: conv::primitive_to_gl_primitive(desc.input_assembler.primitive),
+                    patch_size
```
Comment (msiglreith) [inline src/backend/gl/src/device.rs]: nit: trailing comma

**Comment by grovesNL** [comment]
Alright, added to cache and fixed match/commas (it would be nice to use `rustfmt` at some point because I always miss these commas 😃).

File: src/backend/gl/src/command.rs [hunk 6]
```diff
@@ -478,8 +482,22 @@ impl command::RawCommandBuffer<Backend> for RawCommandBuffer {
...
+            primitive,
+            patch_size,
+            ..
+        } = pipeline;
+
+        if self.cache.primitive != Some(primitive) {
+            self.cache.primitive = Some(primitive);
+        }
+
+        if self.cache.patch_size != patch_size {
```
Comment (msiglreith) [inline src/backend/gl/src/command.rs]: `self.cache.patch_size` never gets updated

Comment (grovesNL) [reply src/backend/gl/src/command.rs]: Whoops :sweat_smile: Fixed

**Pull request merged**